systemtest/edu/washington/escience/myria/systemtest/ParallelIngestS3Test.java

```diff
@@ -0,0 +1,149 @@
+      Type.STRING_TYPE, "d_holidayfl", Type.STRING_TYPE, "d_weekdayfl", Type.STRING_TYPE);
+
+  Schema customerSchema = Schema.ofFields("c_custkey", Type.LONG_TYPE, "c_name", Type.STRING_TYPE, "c_address",
+      Type.STRING_TYPE, "c_city", Type.STRING_TYPE, "c_nation_prefix", Type.STRING_TYPE, "c_nation", Type.STRING_TYPE,
+      "c_region", Type.STRING_TYPE, "c_phone", Type.STRING_TYPE, "c_mktsegment", Type.STRING_TYPE);
+
+  String dateTableAddress = "s3://myria-test/dateOUT.csv";
+  String customerTableAddress = "s3://myria-test/customerOUT.txt";
+
+  @Test
+  public void parallelIngestTest() throws Exception {
+    RelationKey relationKey = RelationKey.of("public", "adhoc", "testParallel");
+
+    Map<Integer, RootOperator[]> workerPlansParallelIngest = new HashMap<Integer, RootOperator[]>();
+    int workerCounterID = 1;
+    for (int workerID : workerIDs) {
+      AmazonS3Source s3Source = new AmazonS3Source(dateTableAddress);
+      CSVFileScanFragment scanFragment =
+          new CSVFileScanFragment(s3Source, dateSchema, workerCounterID, workerIDs.length, '|', null, null, 0);
+      workerPlansParallelIngest.put(workerID, new RootOperator[] { new DbInsert(scanFragment, relationKey, true) });
+      workerCounterID++;
+    }
+    server.submitQueryPlan(new SinkRoot(new EOSSource()), workerPlansParallelIngest).get();
+    assertEquals(2556, server.getDatasetStatus(relationKey).getNumTuples());
+  }
+
+  @Test
+  public void diffParallelIngestTest() throws Exception {
+
+    /* Ingest in parallel */
+    RelationKey relationKeyParallelIngest = RelationKey.of("public", "adhoc", "ingestParallel");
+    Map<Integer, RootOperator[]> workerPlansParallelIngest = new HashMap<Integer, RootOperator[]>();
+    int workerCounterID = 1;
+    for (int workerID : workerIDs) {
+      AmazonS3Source s3Source = new AmazonS3Source(customerTableAddress);
+      CSVFileScanFragment scanFragment =
+          new CSVFileScanFragment(s3Source, customerSchema, workerCounterID, workerIDs.length, ',', null, null, 0);
+      workerPlansParallelIngest.put(workerID, new RootOperator[] { new DbInsert(scanFragment,
+          relationKeyParallelIngest, true) });
+      workerCounterID++;
+    }
+    server.submitQueryPlan(new SinkRoot(new EOSSource()), workerPlansParallelIngest).get();
+    assertEquals(300000, server.getDatasetStatus(relationKeyParallelIngest).getNumTuples());
+
+    /* Ingest the through the coordinator */
+    RelationKey relationKeyCoordinatorIngest = RelationKey.of("public", "adhoc", "ingestCoordinator");
+    server.ingestDataset(relationKeyCoordinatorIngest, server.getAliveWorkers(), null, new FileScan(new UriSource(
+        customerTableAddress), customerSchema, ',', null, null, 0), new RoundRobinPartitionFunction(workerIDs.length));
+    assertEquals(300000, server.getDatasetStatus(relationKeyCoordinatorIngest).getNumTuples());
+
+    /* do the diff at the first worker */
```

Comment:
Per discussion, we'll insert a shuffle with `WholeTupleHashPartitionFunction` before `Difference` so the relations to compare can be partitioned.